### PR TITLE
Added game!

### PR DESCRIPTION
**Pull request by Neel-Nanda**

### CODE REVIEW

File: game.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        }
+      }
+    }
+
+    //This point can only be reached if no one has won and the grid is full - a
+    //stalemate
+    return 3;
+  }
+
+public:
```
Comment (sathiyan-sivathas) [inline game.cpp:191]: This is the second time you've used the public keyword in this class - it would be better to avoid that. I think you've needed to do it because `move` makes use of some of the private methods in this class so needs to be defined after them.

The way we solve this problem is to define the class's methods and member variables in a header file and then implement the methods in a .cpp file. That way the ordering of the functions in the .cpp file isn't important because you've included the header file and will always be able to call a function even if it is defined later in the file.

File: game.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+
+class Grid
+{
+public:
+  Grid() {
+    row=3;
+    col=3;
+    player=1;
+    line_length=3;
```
Comment (sathiyan-sivathas) [inline game.cpp:14]: Normally we'd initialize member variables in an initialization list rather than in the body of the constructor e.g.

```
Grid() :
  row(3),
  col(3)
```

File: game.cpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+#include <iostream>
+#include <vector>
+using namespace std;
+
+
+class Grid
+{
+public:
+  Grid() {
+    row=3;
```
Comment (sathiyan-sivathas) [inline game.cpp:11]: If you wanted to make this function extensible to an arbitrarily sized grid, you could pass in row and column into the constructor so that the calling code can ask for any sized grid.

File: game.cpp (hunk 2 above)
Comment (sathiyan-sivathas) [inline game.cpp:14]: There are three variables here `row`, `col` and `line_length` that are initialized as 3 and never changed by the program. Really, all you're representing here is the size of the grid, so it would be clearer to have a single variable like `grid_size` or something similar.

As above, you might want to pass this variable into the constructor since it would let you play nxn noughts and crosses.

File: game.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    col=3;
+    player=1;
+    line_length=3;
+    // 0 means empty cell, X=1, O=2
+    board=vector<vector<int> > (row, vector<int> (col, 0));
+  }
+
+  virtual ~Grid(){}
+
+  int get_player(){
```
Comment (sathiyan-sivathas) [inline game.cpp:21]: We normally put the opening brace here on the next line. Same goes for `if` and `for` statements.

File: game.cpp [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    return player;
+  }
+
+private:
+  vector<vector<int> > board;
+  int row, col;
```
Comment (sathiyan-sivathas) [inline game.cpp:27]: You can start your member variables with `_` to make it easier to distinguish them from local variables.

File: game.cpp [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+      }
+    }
+    //This point is only reached if every element was the same and non-zero
+    return first_cell;
+  }
+
+  int victory() {
+    //Check vertical and horizontal lines
+    int line_val;
+    for (int r=0; r<row-line_length+1; r++){
```
Comment (sathiyan-sivathas) [inline game.cpp:128]: Following the comment higher up, `row` and `line_length` always have the same value of 3, so this loop isn't doing anything except setting `r=0`. But that's all you need right? This block is checking the horizontal lines, so you don't need to start at different horizontal positions, just the leftmost one. That means you can remove this for loop altogether.

File: game.cpp [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        line_val=check_line(r, c, 1, 0);
+        if (line_val>0)
+        {
+          return line_val;
+        }
+      }
+    }
+
+    for (int r=0; r<row; r++){
+      for (int c = 0; c < col-line_length+1; ++c)
```
Comment (sathiyan-sivathas) [inline game.cpp:140]: As above `c=0` is the only value used here. You can remove this for loop.

File: game.cpp [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  int row, col;
+  //1 is X and goes first, 2 is O
+  int player;
+  int line_length;
+
+  void set_cell(int r, int c, int value)
+  {
+    //If (r, c) is a valid cell, sets it to value
+    //Otherwise, logs an error
+    if (0<=r and r<row and 0<=c and c<col){
```
Comment (sathiyan-sivathas) [inline game.cpp:36]: I had no idea you could use `and` in C++! We normally use `&&`.

File: game.cpp [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        if (line_val>0)
+        {
+          return line_val;
+        }
+      }
+    }
+
+    //Check diagonal lines
+    for (int r=0; r<row-line_length+1; r++){
+      for (int c = 0; c < col-line_length+1; ++c)
```
Comment (sathiyan-sivathas) [inline game.cpp:152]: Similar comments to above - these for loops aren't doing anything. There are only two diagonals you need to check so there isn't a need for all of these loops. You just need to do the diagonal starting at (0,0) and the diagonal starting at (0,2).

**Review by sathiyan-sivathas — COMMENTED** [review]
Hi Neel, I've reviewed your code. It looks good and I think the logic you've used here is sound. You've also commented your code well - it was easy to read the comments and understand what you were trying to do. I've left various comments inline, but my main feedback is on:
- Code structure: you've written all your code in one file and there isn't any real structure to it which makes it a bit hard to follow. Defining a header file that contains definitions of all the functions and member variables would help with this since it allows someone to read the header file to get an idea of what your class does, then read the .cpp file to see the details.
- Overcomplication: there are a few places where your code is more complicated than it needs to be. I'm thinking about some of the unnecessary variables and for loops.

I'll pass this over to Alex for him to add his feedback.

File: game.cpp (hunk 2 above)
Comment (AlexHockey) [reply game.cpp:14]: This isn't just what we'd normally do - it's idiomatic C++.

EDIT: This comment got displayed weirdly. It was in response to Sathiyan's comment about using initialization lists. Definitely use initialization lists!

File: game.cpp [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    if (0<=r and r<row and 0<=c and c<col){
+      board[r][c]=value;
+    }
+    else {
+      cerr<<"Attempt to set invalid cell ("<<r<<", "<<c<<") to "<<value<<endl;
+    }
+  }
+
+  int find_cell(int cell_num) {
+    // The player inputs a number from 1 to 9, corresponding to a cell:
```
Comment (AlexHockey) [inline game.cpp:45]: Comment is inconsistent with the code (and itself!)

File: game.cpp [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+      {
+        cout<<"-----\n";
+      }
+    }
+  }
+
+  int get(int r, int c){
+    //Returns the cell (r, c)
+    //If these are invalid coordinates, logs an error and returns -1
+    if (0<=r and r<3 and 0<=c and c<3){
```
Comment (AlexHockey) [inline game.cpp:94]: Generally you've been using `row` and `column` as the sizes of your grid, but here you are using a hardcoded `3`.

File: game.cpp [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+      if (i<row-1)
+      {
+        cout<<"-----\n";
+      }
+    }
+  }
+
+  int get(int r, int c){
+    //Returns the cell (r, c)
+    //If these are invalid coordinates, logs an error and returns -1
```
Comment (AlexHockey) [inline game.cpp:93]: Comment does not match the code.

File: game.cpp [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    //3 if it's a stalemate, -1 if the cell is occupied and -2 if the cell
+    //number is out of bounds
+    if (0>cell_num or row*col<=cell_num){
+      return -2;
+    }
+    else if (find_cell(cell_num)==0){
+      set_cell(cell_num/row, cell_num%col, player);
+      //cout<<cell_num/row<<" "<<cell_num%col<<endl;
+      //Cycles between 1 and 2
+      player=1+(player)%2;
```
Comment (AlexHockey) [inline game.cpp:206]: Having a separate `switch_player` method might be clearer.

File: game.cpp [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+int main(int argc, char *argv[])
+{
+  Grid board;
+  //-1 if previous move failed, 0 if previous move was a success, 1 or 2 if the
+  //corresponding player won, 3 if it's a stalemate
+  int move_success=0;
+  int cell_num;
+  cout<<"Welcome to Noughts and Crosses!"<<endl;
+  cout<<"Cells are labelled as follows:"<<endl;
+  cout<<"0|1|2\n-+-+-\n3|4|5\n-+-+-\n6|7|8"<<endl;
```
Comment (AlexHockey) [inline game.cpp:229]: You could split the string onto multiple lines to make it easier to read.

**Review by AlexHockey — COMMENTED** [review]
Generally this was very good. I thought the code was broken down into sensible methods, and you commented it well. I think this would be fairly easy to maintain. 

I do have a few high level comments below, and a few more detailed comments inline. 

* A lot of you code uses integers to represent states. For example a cell can either be empty, X , or 0. Or the game can either be "in progress", "player 1 victory", "player 2 victory" or "stalemate". However it is quite difficult to follow what values mean what states. This is a classic example of when to use an enum. 
* You've dealt with the user entering an invalid cell index which is great.  However exactly where you deal with this is a bit inconsistent. For example you check the cell index in both `move()` and `set_cell`. Also the arguments you pass into to functions is a bit inconsistent - e.g. `set_cell` takes a coordinate, but `find_cell` takes a cell number. Generally its best to check input towards the edge of the program and have the internals assume they are being called correctly. So I would probably have `move` validate the cell number passed in is correct, and have all the other functions take coordinate pairs, that they assume are valid.
* Relatedly I wonder if it would have been good to create a simple `Coordinate` class, where the only way to get a coordinate instance is by calling a `cell_index_to_coordinate` method that checks the index is valid. Having a coordinate class might also have made the check line method a little bit easier to read (as it would only have two arguments not four). 
* Don't be quite so worried about brevity of function names. For example it's not clear what `get` is doing just from the name, but I would definitely know what `get_cell_state` was doing.
